### ISSUE TITLE
[Bug] Conditional product dependencies `.when(platforms:)` were wrongly detected

### ISSUE THREAD

**Issue by Stannieman** [trinhngocthuyen/cocoapods-spm#113]
### What happened?

Hi! Not sure if this is a bug or some unusual configuration on our end.
When adding this to our project, as soon as the dependency is added to a podspec file, the pods project gets corrupted and opening results in the error below. Any idea what could cause this?

> The project ‘Pods’ is damaged and cannot be opened. Examine the project file for invalid edits or unresolved source control conflicts.

> -[XCSwiftPackageProductDependency _setSavedArchiveVersion:]: unrecognized selector sent to instance 0x60002ca42100

**Comment by trinhngocthuyen** [comment]
Thanks for reaching out! Could you please provide more info for troubleshooting, for example:
- What packages were being used?
- Does pod install finish successfully?
- ...

I would very much appreciate if we have a demo project or Podfile for reproduction. Thank you!

**Comment by Stannieman** [comment]
Yeah I could see this question coming. Unfortunately our project structure is a bit chaotic at the moment 😅
Let me first clean that up a bit and try to come up with a a more minimal repro before I share anything there.
Just to be sure: Are cocoapods-compact-spec and cocoapods-xcconfig-hooks plugins (or any other gems for that matter) needed for this to work or should it work without them too?

**Comment by trinhngocthuyen** [comment]
cocoapods-compact-spec and cocoapods-xcconfig-hooks and this plugin are independent of each other.
they are just to reduce some declaration effort (in podspecs) for the example project.

**Comment by Stannieman** [comment]
Ok so after some cleanup in our podspecs and Podfile the Pods project does not become corrupted anymore. That's already a big win.
Now I encounter a new issue. When using an SPM package that itself has other dependencies it seems these transitive dependencies are not correctly resolved.

An example is TOMLKit.
![image](https://github.com/user-attachments/assets/250beaec-d4ac-4070-a28f-59ea83e764f0)
Even when adding CTOML to the list of producs in the Podfile and adding `s.spm_dependency "TOMLKit/CTOML"` to the podspec it cannot find it.

Is this supposed to work?

**Comment by trinhngocthuyen** [comment]
Glad to hear that Pods project becomes normal.
For transitive dependencies, it is supposed to work. I have been fixing quite some issues like this, especially with non-Swift packages.
If it's ok to share some more info  such as the package declaration in Podfile, I can help take a look

**Comment by Stannieman** [comment]
TOMLKit was just a simpler example. The one I actually need is mqtt-nio.
```
spm_pkg 'mqtt-nio',
    :url => 'https://github.com/swift-server-community/mqtt-nio.git',
    :version => '2.11.0',
    :products => ['MQTTNIO']
```
and in some podspec of our project (that lives in the same repo) `s.spm_dependency 'mqtt-nio/MQTTNIO'`.

I'll ask if it is ok to share the entire Podfile and the podspec of the pod where I need mqtt-nio.

**Comment by Stannieman** [comment]
Ok brace yourself, here is our huge podfile. Names of our components (that are pods in the same repo), apps and company are redacted.

```
plugin 'cocoapods-developing-folder'
plugin 'cocoapods-spm'

# Settings
Pod::IPHONEOS_DEPLOYMENT_TARGET = '16.6'
Pod::MACOS_DEPLOYMENT_TARGET = '14.0'

use_frameworks!
install! 'cocoapods', :deterministic_uuids => false

# Specs repositories
source 'https://github.com/CocoaPods/Specs.git'
project 'MyApp.xcodeproj'

def external_production_dependencies
  firebase_version = '11.1.0'

  pod 'MyPod1', :podspec => './ExternalPodspecs/MyPod1.podspec'
  pod 'AnyCodable-FlightSchool', '~> 0.6.4'
  pod 'MyPod2', :git => 'REDACTED', :tag => '1.0.3'
  pod 'BetterCodable', '~> 0.4.0'
  pod 'boost-iosx-headers', :podspec => './ExternalPodspecs/boost-iosx-headers.podspec'
  pod 'DeviceKit'
  pod 'DGCharts', '5.1.0'
  pod 'EasyPeasy', '1.10.0'
  pod 'MyPod3', :podspec => './ExternalPodspecs/MyPod3.podspec'
  pod 'FirebaseAnalytics', firebase_version
  pod 'FirebaseCore', firebase_version
  pod 'FirebaseCrashlytics', firebase_version
  pod 'FirebaseMessaging', firebase_version
  pod 'GoogleUtilities', '8.0.2'
  pod 'MyPod4', :git => 'REDACTED', :tag => '0.32.0'
  pod 'KMPlaceholderTextView', '1.4.0'
  pod 'ObjectMapper', '4.4.2'
  pod 'MyPod5', :podspec => './ExternalPodspecs/MyPod5.podspec'
  pod 'PSPDFKit', podspec: 'https://my.pspdfkit.com/pspdfkit-ios/13.8.0.podspec'
  pod 'Realm', '10.52.2'
  pod 'RealmSwift'
  pod 'RxCocoa'
  pod 'RxGesture'
  pod 'RxSwift', '6.7.1'
  pod 'SkyPath', '2.2.3'
  pod 'SplitView', :podspec => './ExternalPodspecs/SplitView.podspec'
  pod 'SwiftyDraw', '~> 2.4.1'
  pod 'SwiftyJSON', '5.0.2'
  pod 'Swinject', '2.9.1'
  pod 'SwinjectAutoregistration', '2.9.1'
  pod 'URITemplate', '~> 3.0.1'
  pod 'MyPod6', :podspec => './ExternalPodspecs/MyPod6.podspec'
  pod 'MyPod7', :podspec => './ExternalPodspecs/MyPod7.podspec'
  pod 'WhirlyGlobe', :podspec => './ExternalPodspecs/WhirlyGlobe.podspec'
  # This Zip version identifies itself as 2.1.2, but it is actually 2.1.2 + 1 additional commit that fixes https://github.com/marmelroy/Zip/issues/245
  pod 'Zip', :git => 'https://github.com/marmelroy/Zip.git', :commit => 'bca30f6d6c7d37cbc4aa8f6b0002e281dcc36195'

  spm_pkg "mqtt-nio",
    :url => 'https://github.com/swift-server-community/mqtt-nio.git',
    :version => '2.11.0',
    :products => ['MQTTNIO']
end

def production_components
  pod 'Component1', :path => './Components/Component1'
  pod 'Component2', :path => './Components/Component2'
  pod 'Component3', :path => './Components/Component3'
  pod 'Component4', :path => './Components/Component4'
  pod 'Component5', :path => './Components/Component5'
  pod 'Component6', :path => './Components/Component6'
  pod 'Component7', :path => './Components/Component7'
  pod 'Component8', :path => './Components/Component8'
  pod 'Component9', :path => './Components/Component9'
  pod 'Component10', :path => './Components/Component10'
  pod 'Component11', :path => './Components/Component11'
  pod 'Component12', :path => './Components/Component12'
  pod 'Component13', :path => './Components/Component13'
  pod 'Component14', :path => './Components/Component14'
  pod 'Component15', :path => './Components/Component15'
  pod 'Component16', :path => './Components/Component16'
  pod 'Component17', :path => './Components/Component17'
  pod 'Component18', :path => './Components/Component18'
  pod 'Component19', :path => './Components/Component19'
  pod 'Component20', :path => './Components/Component20'
  pod 'Component21', :path => './Components/Component21'
  pod 'Component22', :path => './Components/Component22'
  pod 'Component23', :path => './Components/Component23'
  pod 'Component24', :path => './Components/Component24'
  pod 'Component25', :path => './Components/Component25'
  pod 'Component26', :path => './Components/Component26'
  pod 'Component27', :path => './Components/Component27'
  pod 'Component28', :path => './Components/Component28'
  pod 'Component29/Subspec', :path => 'Components/Component29'
  pod 'Component30', :path => './Components/Component30'
  pod 'Component31', :path => './Components/Component31'
  pod 'Component32', :path => './Components/Component32'
  pod 'Component33', :path => './Components/Component33'
  pod 'Component34', :path => './Components/Component34'
  pod 'Component35', :path => './Components/Component35'
  pod 'Component36', :path => './Components/Component36'
  pod 'Component37', :path => './Components/Component37'
  pod 'Component38', :path => './Components/Component38'
  pod 'Component39', :path => './Components/Component39'
end

def external_test_dependencies
  pod 'Nimble', '13.4.0'
  pod 'OHHTTPStubs/Swift', '9.1.0'
  pod 'Quick', '7.6.2'
  pod 'SwiftFormat/CLI', '0.54.3'
  pod 'SwiftLint', '0.56.2'
  pod 'RxBlocking'
  pod 'RxTest'
  pod 'SnapshotTesting', :podspec => './ExternalPodspecs/SnapshotTesting.podspec'
  pod 'Stencil'
  pod 'SwiftyMocky', :git => 'https://github.com/MakeAWishFoundation/SwiftyMocky.git', :tag => '4.2.0'
end

def test_components
  pod 'Component1/TestHelpers', :path => './Components/Component1'
  pod 'Component7/TestHelpers', :path => './Components/Component7'
  pod 'Component8/TestHelpers', :path => './Components/Component8'
  pod 'Component10/TestHelpers', :path => './Components/Component10'
  pod 'Component11/TestHelpers', :path => './Components/Component11'
  pod 'Component12/TestHelpers', :path => './Components/Component12'
  pod 'MyPod2/TestHelpers', :git => 'REDACTED', :tag => '1.0.3'
  pod 'Component18Test', path: './Components/Component18'
  pod 'Component19/TestHelpers', :path => './Components/Component19'
end

target 'MyAppVariant1' do
  platform :ios, IPHONEOS_DEPLOYMENT_TARGET
  # Pods
  external_production_dependencies
  production_components
end

target 'MyAppVariant2' do
  platform :ios, IPHONEOS_DEPLOYMENT_TARGET
  # Pods
  external_production_dependencies
  production_components
end

target 'MyAppVariant3' do
  platform :ios, IPHONEOS_DEPLOYMENT_TARGET
  # Pods
  external_production_dependencies
  production_components
end

target 'DesignSystemShowcase' do
  platform :ios, IPHONEOS_DEPLOYMENT_TARGET
  # Pods
  pod 'Component5', :path => './Components/Component5'
  pod 'Component8', :path => './Components/Component8'
  pod 'SwiftGen', '~> 6.0'
end

target 'Component1CommandLineTestTool' do
  platform :osx, MACOS_DEPLOYMENT_TARGET
  # Pods
  pod 'Component1', :path => './Components/Component1'
  pod 'Component7', :path => './Components/Component7'
end

target 'AllTests' do
  platform :ios, IPHONEOS_DEPLOYMENT_TARGET
  # Pods
  external_production_dependencies
  production_components
  external_test_dependencies
  test_components
end

# Post install
post_install do |installer|
  installer.pods_project.targets.each do |target|
    target.build_configurations.each do |config|

      # We have a bunch of third party pods that have macOS deployment target set to < 12.
      # When building with Xcode 15+ they actually require macOS 12+.
      # They (unknowingly?) depend on libarclite which from Xcode 15 requires macOS 12. When their CI is Xcode 14 they will never notice.
      # To fix this we bump those deployment targets here.
      if config.build_settings['MACOSX_DEPLOYMENT_TARGET'] &&
        Gem::Version.new(config.build_settings['MACOSX_DEPLOYMENT_TARGET']) < Gem::Version.new(MACOS_DEPLOYMENT_TARGET)
        config.build_settings['MACOSX_DEPLOYMENT_TARGET'] = MACOS_DEPLOYMENT_TARGET
      end

      config.build_settings['ARCHS'] = 'arm64'
      # Bitcode disabling is needed for MyPod4 pod
      config.build_settings['ENABLE_BITCODE'] = 'NO'
      config.build_settings['SWIFT_OPTIMIZATION_LEVEL'] = '-O'
      if !(['Component4', 'SBTUITestTunnelCommon'].include?(target.name) || target.name.start_with?("Firebase"))
        config.build_settings['SWIFT_INSTALL_OBJC_HEADER'] = 'NO'
      end

      if config.name == 'Debug'
        config.build_settings['SWIFT_OPTIMIZATION_LEVEL'] = '-Onone'
        config.build_settings['DEBUG_INFORMATION_FORMAT'] = 'dwarf'
      end

      # GEOSwift exception is necessary to avoid a 'Symbol not found' crash: https://forums.developer.apple.com/forums/thread/687259
      unless target.name == 'GEOSwift'
          config.build_settings["IPHONEOS_DEPLOYMENT_TARGET"] = IPHONEOS_DEPLOYMENT_TARGET
      end
      if (['Component5', 'Component4', 'Component7.common-iOS', 'Component7.common-macOS', 'Component7.common-Tests',
          'Component10.common', 'Component10.common-Tests'].include?(target.name))
        config.build_settings["ASSETCATALOG_COMPILER_GENERATE_SWIFT_ASSET_SYMBOL_EXTENSIONS"] = 'YES'
      end
    end
    if target.respond_to?(:product_type) and target.product_type == "com.apple.product-type.bundle"
      target.build_configurations.each do |config|
          config.build_settings['CODE_SIGNING_ALLOWED'] = 'NO'
      end
    end
  end

end

# Post integrate
post_integrate do |installer|
  # Because of WhirlyGlobe, we need to enable GPU flag to disable Metal validation
  scheme = Xcodeproj::XCScheme.new("MyApp.xcodeproj/xcshareddata/xcschemes/MyAppVariant1.xcscheme")
  scheme.launch_action.xml_element.attributes['enableGPUValidationMode'] = "1"
  scheme.save!
end

# Ignore warnings in 3th party pods. An external gem is used for this because the native Cocoapods
# solution doesn't work in our case. More information on this can be found on this url:
# https://github.com/leavez/cocoapods-developing-folder/blob/master/README.md#-inhibit-warnings-for-specific-pods
#
# Some pods are maintained by us, but these warnings should be fixed in their repo's.
# - MyPod4
# - MyPod7
inhibit_warnings_with_condition do |pod_name, pod_target|
  %w(
     Realm
     RxSwift
     RxCocoa
     EasyPeasy
     PSPDFKit
     Maaku
     GoogleUtilities
     FirebaseCore
     libcmark_gfm
     Quick
     SnapshotTesting
     SwinjectAutoregistration
     FirebaseCrashlytics
     FirebaseAnalytics
     MyPod4
     MyPod7
     MyPod3
     GEOSwift
     ObjectMapper
     SwiftyJSON
     BetterCodable
     ).include? pod_name
end
```

And the podspec of the pod where I want to use mqtt-nio:
```
Pod::Spec.new do |s|
  s.name             = 'Component1'
  s.version          = '1.0.0'
  s.summary          = 'Summary'
  s.description      = <<-DESC
  Decription
  DESC
  s.homepage      = "MyURL"
  s.license       = 'MyCompany'
  s.author        = { "MyCompany" => "REDACTED" }
  s.swift_version = '5.9'
  s.source        = { :git => 'REDACTED', :tag => s.version }

  s.default_subspecs = 'Core/Classes'

  s.subspec 'Core' do |cs|
    cs.subspec 'Classes' do |ls|
      ls.source_files = ['Component1/Core/Classes/**/*.swift']
    end
  end

  # This subspec should only be used in test targets. You can add this
  # subspec where you want to use the shared generated mocks or other
  # test helpers.
  s.subspec 'Tests' do |mocks|
    mocks.source_files = ['Component1/Tests/**/*.swift']
    mocks.dependency 'SwiftyMocky'
  end

  # Dependencies
  s.dependency 'Component7'
  s.spm_dependency 'mqtt-nio/MQTTNIO'
end
```

**Comment by trinhngocthuyen** [comment]
Thanks for the generous info haha.

I've figured out the root cause of MQTTNIO.
The plugin mist-detects NIOSSL as a dependency MQTTNIO. But this NIOSSL is meant for Linux/MacOS platform only
https://github.com/swift-server-community/mqtt-nio/blob/deae8ea/Package.swift#L30
<img width="804" alt="Screenshot 2024-10-10 at 22 09 38" src="https://github.com/user-attachments/assets/5c0d8cb3-938c-4e65-9e36-a97b69433c45">
I'll work on a fix for this case.

**Comment by trinhngocthuyen** [comment]
Hi there. I've made a fix for conditional product dependencies in #115.
Kindly check it out. Thank you!

---

On a side note about TOMLKit. I saw you used `s.spm_dependency "TOMLKit/CTOML"` and it's not quite correct. The package declaration only exposes the product `TOMLKit` (not `CTOML`) https://github.com/LebJe/TOMLKit/blob/main/Package.swift#L14.
So it should be `s.spm_dependency "TOMLKit/TOMLKit"` or simply `s.spm_dependency "TOMLKit"`

**Comment by Stannieman** [comment]
Thanks for the quick action!
Pod install crashed due to platform being null, but I fixed it by changing that line in podfile.rb to
`@platforms ||= target_definition_list.map { |d| d.platform&.name }.compact.uniq || [:ios]`.
I don't know if that's an ok thing to do or it can cause problems though.

After that AND removing the Component1CommandLineTestTool target (macOS) in the Podfile (so only iOS targets are left) it built succesfully!
Without removing the macOS target I get a `missing required module 'CNIOAtomics'` error.
Also, the AllTests target fails with that same error regardless.

I suspect the issue is that, when both iOS and macOS targets exist together, these pod targets get suffixed with -iOS and -macOS and the plugin does not handle this.
E.g. instead of Component1 there will be Component1-iOS and Component1-macOS.
The story with AllTests is likely the same, there Component1.default-Tests is used.

**Comment by trinhngocthuyen** [comment]
I made a fix in #116. Could you help check again? Thank you!

**Comment by Stannieman** [comment]
I don't know why, but some of my local pods are also named SomePod.common-iOS, and the pod in which I use mqtt-nio also has a SomePod.default-Tests target (probably due to Tests subsped). For these targets the dependencies are still not added. I had more luck with using target.product-name for looking up in the dictionary.
